### ISSUE TITLE
[Requestor enchancement] Make it possible to rent only x out of y cores on a provider and leave the remaining for the rest of the network

### ISSUE THREAD

**Issue by cryptobench** [golemfactory/yagna#2356]
Let's say you need 2 cores for a specific service and you have a market strategy that selects the absolute cheapest provider on the network within these requirements. It now turns out that a machine with 10 cores is the cheapest one on the network, but you only need 2. Because it's the cheapest then naturally you select that provider, but now the remaining 8 cores are just stuck idling. 

There's a lot of issues with this because:

1) The network is not utilised in a proper efficient way
2) The provider now have 8 cores that are not actively billed for because no cpu-time is spent by the requestor. This is effectively decreasing the providers earnings, because he could just as-well have someone pay to use the remaining 8 cores.
3) If I want to create a cloud and rent out instances to my customers as a middleman service, then I could very quickly rent out a very high capacity of the network. How this would work is that if a customer of mine wants to rent 2 cores for some service, then I go ahead and create a demand on the network for 2 cores, but all providers with 2 cores are unavailable because they are already computing something and now I have to pick from the providers with 4 cores available. Even though I have rented 4 cores, then I still only want to give access to 2 cores to my customer, so I will just let the 2 others sit idle and do nothing. 
I could of course provision 2 customers wanting 2 cores to the same provider, but it will be tough for me to limit them from actually only using 2 cores. I can of course investigate doing this, but again let's make Golem as easy as possible for our requestors and provide them the tools out of the box so we don't have to create all these hacky solutions.
4) If we look at how clouds currently provision machines to the customers, then they have 1 big machine with let's say 64 cores, and when someone needs 4 cores then it provisions a VM with 4 cores and now we have 60 cores left to rent out to other customers on this specific machine. Golem should have the same ideology about its providers on the network.
5) I am a requestor and I don't want to pay for example **THREE** providers when it could be run on a single provider for **atleast** 1/3rd of the cost. Money is the most important factor of gaining requestors on the network.

TL:DR - I believe the vm ExeUnit should be able to spawn with a specific amount of cores requested by the requestor and not all cores that the provider is advertising on the network. This also means that a provider should be able to create multiple agreements and advertise the cores it has left automatically to the network.

**Comment by evik42** [comment]
This becomes more complicated with #2357 where one VM's network usage can have a direct effect on the bandwidth available for another VM running on the same host.

**Comment by cryptobench** [comment]
> This becomes more complicated with #2357 where one VM's network usage can have a direct effect on the bandwidth available for another VM running on the same host.

A lot of software for example in the torrenting world allow users to limit the amount of bandwidth the application can use. Couldn't we implement something similar here?

Here's an idea:

We have a requestor browsing the network providers individual offers. When looking at the offers the provider will advertise the following: `download : 120 , upload: 300`

So this is the theoretical max speeds that the provider has available and can rent to the network. A requestor should then be able to specify when renting out a machine the upload/download speed they wish to allocate to their providers VM. Now the ExeUnit should provision a VM with the network speeds limited to the requestors needs. Next what happens is that it deducts these values from the providers offer template and advertise the leftover capacity instead. 

So if we rented 90 mbit out of 100 mbit, then the providers offer template should advertise it has 10 mbit available.



The proposed solution above definitely has some caveats for providers. In the case that for example a provider rents out 10 out of 20 cores available, and the requestor also rents out ALL the bandwidth of the provider, then it has 10 cores left and 0/0 bandwidth left for other requestors. So the question is what to do here? We could instead take another approach to all of this where users can choose between:

option 1) which is the first one i described where it deducts from the total capacity based on whats rented already
option 2) We can introduce something like vSpeeds, which is a shared link between all requestors renting something on a provider. This is similar to renting a cheap VPS where the cloud provider advertises 4 vCPU's, which effectively means that the machine is overprovisioned and you're sharing it with other customers.

**Comment by konradbjk** [comment]
@broadcastmonkey it has been 4 months, and it is still missing. I believe if we would implement even the wrong analogy, we would be better now.

What people expect to buy is # of vCPU (as this is well-known logic on the market). You can propose a better option than vCPU, but it will need to be additionally clarified for people. We are not in a stage where we can re-educate users, as we do not have a late majority, let alone a proper market-fit. 

@evik42 could you please give us some estimation on how many tasks would be impacted (really utilized the 80%+ of network bandwidth of a provider) via this issue - looking at the last, let's say 90 days? Maybe it is a characteristic of a task and we can fix it using the custom market strategy for such tasks. Then I personally, do not see it as a problem anymore

**Comment by konradbjk** [comment]
> “If you can not measure it, it does not exist.”
I like the above quote regarding problems:p 

We are community-driven, AFAIK it is something they have raised. 

Think like a provider @evik42. We run AMD EPYC, which costs like $3000 alone (not to mention electricity), and we are getting paid by running a to-do app on it. It utilizes maybe one core and almost no computation for an hour - we get 1 GLM for that. We could easily earn 16 times that or run 16 of those apps...

So what do providers do? They overprovision and run 100 providers on one machine. We use the number of providers as one of our metrics, and due to this issue, we cannot rely on this metric.

@cryptobench did I miss anything?

**Comment by evik42** [comment]
@konradbjk I think you can make a valid use case for most features, but it is a question of priorities and ROI. Solving this issue does not improve the current state of the network. Your example is great in case there are demands that the provider had to reject because it was occupied, which is currently rarely the case. (Also I think your example of someone building a provider on AMD EPYC is wishful thinking given the current earning potential of the network.)

A system with a lot of cores can be better utilized by running multiple providers, you do not need to overprovision, you can run a provider with 4 cores and run as many as your system can fit. Overprovisioning is a behavior independent of the number of cores, currently it provides the highest earning potential because there is no negative consequence for providing sub par performance.

I responded because you tagged me but I do not want to argue about product decisions. If you can create a compelling case where the benefits out weight the work required, it will be prioritized and put on the team's priorities. For me you failed to do that so far but I am not a decision maker in these questions.
If you need any data about our network then please check with the devops team.

**Comment by broadcastmonkey** [comment]
@konradbjk As @evik42 stated for the time being there are more important and relevant issues to be resolved that will have higher impact on the Golem ecosystem. Having said that this issue is valid and we will need to get back to it at some point. 

Best

**Comment by konradbjk** [comment]
What I did there is triage and some research that we need to include in other cases. Maybe there are more important issues, maybe not - we are not so sure because there is no universal rating of issues. Within those 270+ issues, we might also have duplicates (one solution could solve multiple raised problems or multiple solutions were presented to solve one problem). Obviously, things related to network stability and its status are more important - maybe those should not if you think about a startup, you need to produce a ton of features while the previous ones are in bad shape. This is Tomek's decision.

@evik42 regarding AMD EPYC, check https://stats.golem.network/network/providers/online and look for providers with 24+ cores. The list is long.
...or go to https://api.stats.golem.network/v1/network/online and search for specific - I have found 14 EPYC providers out of 500-ish.

**Comment by cryptobench** [comment]
Our biggest issue is the lack of requestors on the network and I think the following is one of the core issues that needs to be addressed:

I am a requestor and I don't want to pay for example THREE providers when it could be run on a single provider for atleast 1/3rd of the cost. Money is the most important factor of gaining requestors on the network.

Golem is utilizing Docker images as its base for GVMI - this usually means that we are into microservices realm. I as a requestor are looking to run my backend with 3 replicas on the network. I know for a fact that they don't need anything more than 1 GB of RAM + 1 vCPU per replica. 

Let's now examine my options:

1) Spin up the backend on 3 different providers and pay for possibly 3 whole machines that are much larger than my needs. Which also results in me possibly squatting a more powerful machine than my needs which other people could have use of. I also don't really have control over spendings in the case that suddenly some functions runs unexpectedly and hogs 4 vCPU's instead of the expected 1 vCPU. There's no limits here for how much I can use, so what happens just happens.

2) I could spin up all 3 replicas on the same provider to save on costs but this doesn't really solve my need for high availability. There's also a chance that the provider will be resource constrained if there's a spike in all 3 replicas. I also have to implemented my own loadbalancing solution here to distribute the traffic between the replicas on the same host.

3) If it was possible I could specify the resource requests for my images which would deduct the requested resources from a randomly selected providers capacity on the network. This would result in me not squatting providers from other people, it would also result in the requestor being able to more easily calculate their possible spendings as there's no chance that you're using more than that 1 vCPU you rented. There's no sudden spikes that might result in 4 vCPU's being used because its simply not possible.

**Comment by cryptobench** [comment]
Providers are also talking about this issue here https://discord.com/channels/684703559954333727/773872812091768852/1103989136249065492

**Comment by krunch3r76** [comment]
This is a compelling idea to maximize utilization of idle resources. I will share my initial thoughts as well on resource provisioning across multiple requestors simultaneously:

### Bandwidth
The bandwidth issue may be solved by throttling the bandwidth on the exeunit. One alternative simple model is that each agreement can have a specific ul/dl bandwidth set equally partitioned so that if all agreements were taken then all would be used up. This would be implemented as a percentage e.g. 5 agreements 20% ul/dl bandwidth. Bandwidth would be determined by iperf or a home brew iperf. [iperf3 homepage](https://iperf.fr/)

While there are ways to throttle for max bandwidth they are not reliable in the userspace (non admin). So this would introduce a new setting, **max bandwidth _percentage_** and be **max number of agreements** (where **_memory per agreement_** is the computed property e.g. 12gib max, 12 agreements, ->1gib memory per agreement

As to how to divide bandwidth, that would require additional logic to be coded into golemsp/yagna, but as cryptobench pointed out, we've seen software that accomplished this before and can model after that.

### Cores
As for how to divide cores, linux provides a standard cpu affinity tool (from util-linux) called taskset [taskset manpage](https://www.man7.org/linux/man-pages/man1/taskset.1.html). This tool can launch exeunit process to help schedule cpu's differentially. This is a hinting system and works under ideal loads (i.e. truly idle systems). The number of cores per agreement would be however many cores can be equally divided into the number of agreements. e.g. max 10 cores allowing 5 agreements -> 2 cores per agreement.

So this introduces a new setting, in addition to **max cores** there would be **max number of agreements** (where **_cores per agreement_** is the computed property and computed similarly as for bandwidth).

### Memory
With this proposed new setting of **max number of agreements** (along with computed max bandwidth percentage and max cores) we also need a **max memory** and computed property of **_max memory per agreement_** equally partitioned as before.

So I don't see at this moment many new problems or complications that would arise from allocating resources (cores and bandwidth AND memory) separately. There is however a possible issue with i/o intensive instances that share the same bus. This I cannot imagine can be solved and would be an implicit drawback on nodes that divide resources for rent multiply.

These are my preliminary thoughts and hope they give some encouragement and help.

**Comment by RoyKas** [comment]
With new features, such as Golem Cache, it is increasingly important to use only the #vcpu that is needed (cache should only need 1) and have remaining resources available to at least one additional activity.
Maybe create 2 groups of activities, like already in the tutorials, services and tasks, and allow 1 of each concurrently. This should cover the most immediate needs and not pose too big a difficulty. Services would typically not require much cores, whereas tasks require as many cores as available.

**Comment by cryptobench** [comment]
Closing as i'm opening a discussion instead https://github.com/golemfactory/yagna/discussions/3005
